File: easy/20._Valid_Parentheses.py
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']',
# determine if the input string is valid.
#
# An input string is valid if:
#
# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
# Every close bracket has a corresponding open bracket of the same type.
#
# Input: s = "()"
# Output: true
#
# Input: s = "()[]{}"
# Output: true
#
# Input: s = "(]"
# Output: false
class Solution:
    def isValid(self, s: str) -> bool:
        '''
        Time complexity O(n)
        Space complexity O(n/2)
        '''
        stack = list()
        opn = set('({[')
        close = set(')}]')
        for el in s:
            if el in opn:
                stack.append(el)
            if el in close:
                if stack:
                    last_el = stack.pop()
                    if el == ']' and last_el == '[':
                        continue
                    elif el == '}' and last_el == '{':
                        continue
                    elif el == ')' and last_el == '(':
                        continue
                return False

        return False if stack else True

File: go/easy/20_valid_parentheses.go
package main

//Given a string s containing just the characters '(', ')', '{', '}', '[' and ']',
//determine if the input string is valid.
//An input string is valid if:
//Open brackets must be closed by the same type of brackets.
//Open brackets must be closed in the correct order.
//Every close bracket has a corresponding open bracket of the same type.
//
//
//Example 1:
//
//Input: s = "()"
//Output: true
//Example 2:
//
//Input: s = "()[]{}"
//Output: true
//Example 3:
//
//Input: s = "(]"
//Output: false

import "fmt"

func isValid(s string) bool {
	if len(s)%2 != 0 || len(s) == 0 {
		return false
	}
	bracketMap := map[rune]rune{
		']': '[',
		'}': '{',
		')': '(',
	}

	stack := []rune{}
	for _, bracket := range s {
		if openBracket, ok := bracketMap[bracket]; ok {
			if len(stack) == 0 || stack[len(stack)-1] != openBracket {
				return false
			} else {
				stack = stack[:len(stack)-1]
			}
		} else {
			stack = append(stack, bracket)
		}
	}
	return len(stack) == 0
}

func main() {
	fmt.Println(isValid("){"))
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
}
